src/Actions/SlotAwareRawAction.php

```diff
@@ -15,36 +15,58 @@
  * @ingroup Actions
  */
 class SlotAwareRawAction extends RawAction {
-    public function __construct(
-        $article,
-        $context
-    ) {
-        $mediaWikiServices = MediaWikiServices::getInstance();
-
-        $parser = $mediaWikiServices->getParser();
-        $permissionManager = $mediaWikiServices->getPermissionManager();
-        $revisionLookup = $mediaWikiServices->getRevisionLookup();
-
-        if ( method_exists( $mediaWikiServices, 'getRestrictionStore' ) ) {
-            $restrictionStore = $mediaWikiServices->getRestrictionStore();
-        } else {
-            $restrictionStore = null;
-        }
-
-        $userFactory = $mediaWikiServices->getUserFactory();
+	public function __construct(
+		$article,
+		$context
+	) {
+		$mediaWikiServices = MediaWikiServices::getInstance();
+
+		$parser = $mediaWikiServices->getParser();
+		$permissionManager = $mediaWikiServices->getPermissionManager();
+		$revisionLookup = $mediaWikiServices->getRevisionLookup();
+		$hookContainer = $mediaWikiServices->getHookContainer();
+
+		if ( method_exists( $mediaWikiServices, 'getRestrictionStore' ) ) {
+			$restrictionStore = $mediaWikiServices->getRestrictionStore();
+		} else {
+			$restrictionStore = null;
+		}
 
-        parent::__construct(
-            $article,
-            $context,
-            $parser,
-            $permissionManager,
-            $revisionLookup,
-            $restrictionStore,
-            $userFactory
-        );
-    }
+		if( version_compare( MW_VERSION, '1.42', '>=' ) ) {
+			$userFactory = $mediaWikiServices->getUserFactory();
+			parent::__construct(
+				$article,
+				$context,
+				$parser,
+				$permissionManager,
+				$revisionLookup,
+				$restrictionStore,
+				$userFactory
+			);
+		} elseif ( version_compare( MW_VERSION, '1.40', '>=' ) ) {
+			parent::__construct(
+				$article,
+				$context,
+				$parser,
+				$permissionManager,
+				$revisionLookup,
+				$restrictionStore
+			);
+		} elseif ( version_compare( MW_VERSION, '1.39', '>=' ) ) {
```

Comment:
```suggestion
		} else {
```

We no longer need to support versions older than 1.39. This should also be reflected in the `extension.json`.